substrate/frame/collective/src/migrations/v5.rs

```diff
@@ -0,0 +1,200 @@
+		);
+		let count = ProposalOf::<T, I>::iter().count();
+		if Proposals::<T, I>::get().len() != count {
+			log::info!("collective proposals count inconsistency");
+		}
+		if Members::<T, I>::get().len() > <T as Config<I>>::MaxMembers::get() as usize {
+			log::info!("collective members exceeds MaxMembers");
+		}
+
+		Ok((count as u32).encode())
+	}
+
+	fn on_runtime_upgrade() -> Weight {
+		let storage_version = StorageVersion::get::<Pallet<T, I>>();
+		log::info!(
+			target: "runtime::collective",
+			"Running migration for collective with storage version {:?}",
+			storage_version,
+		);
+
+		if storage_version <= 4 {
+			let mut weight = T::DbWeight::get().reads(1);
+
+			// ProposalOf
+			for (hash, proposal) in ProposalOf::<T, I>::drain() {
+				weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+				match <T as Config<I>>::Preimages::note(proposal.encode().into()) {
+					Ok(new_hash) => {
+						weight = weight.saturating_add(T::DbWeight::get().writes(1));
+						if new_hash != PreimageHash::from_slice(hash.as_ref()) {
+							log::info!(
+								target: "runtime::collective",
+								"Preimage hash mismatch for proposal {:?}. Expected {:?}, got {:?}",
+								hash,
+								new_hash,
+								PreimageHash::from_slice(hash.as_ref())
+							);
+						}
+						if !<T as Config<I>>::Preimages::is_requested(&new_hash) {
+							log::info!(
+								target: "runtime::collective",
+								"Preimage for proposal {:?} was not requested",
+								hash,
+							);
+						}
+					},
+					Err(_) => {
+						log::info!(
+							target: "runtime::collective",
+							"Failed to note preimage for proposal {:?}",
+							hash,
+						);
+					},
+				}
+			}
+
+			// Proposals
+			Proposals::<T, I>::kill();
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+
+			// Voting
+			for (hash, vote) in Voting::<T, I>::drain() {
+				weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 2));
+				crate::Voting::<T, I>::insert(
+					PreimageHash::from_slice(hash.as_ref()),
+					crate::Votes::<T::AccountId, BlockNumberFor<T>, <T as Config<I>>::MaxMembers> {
+						index: vote.index,
+						threshold: vote.threshold,
+						// the following operations are safe since the bound was previously enforced
+						// by runtime code
+						ayes: BoundedVec::truncate_from(vote.ayes),
+						nays: BoundedVec::truncate_from(vote.nays),
+						end: vote.end,
+					},
+				);
+			}
+
+			// Members
+			crate::Members::<T, I>::put(BoundedVec::truncate_from(Members::<T, I>::get()));
+			weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+			// StorageVersion
+			StorageVersion::new(5).put::<Pallet<T, I>>();
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+
+			weight
+		} else {
+			log::info!(
+				target: "runtime::collective",
+				"Attempted to apply migration to V5 but failed because storage version is {:?}",
+				storage_version,
+			);
+			T::DbWeight::get().reads(1)
+		}
+	}
```

Comment:
There's a lot of nesting and big conditional statements here. I wonder if the code could be simplified a bit?

Reply:
I removed the big if, thanks to the `VersionedMigration`.
Please let me know if something else comes up to your mind that can improve this.